### PR TITLE
Auth caching, attempt 2

### PR DESCRIPTION
**Pull request by rainersigwald**
The last attempt at caching was too general and got backed out in #223. What we really want is by "scope" but we don't have a great way to predict a scope in advance, they're generally given back by the server.

This time, I'm trying to cache by URL (excluding query string), which will be very helpful when hitting chunked upload URLs over and over.

See details in the second commit message especially.

- Enable auth caching by URL
- Quirk for ACR 401 on ".../blobs/uploads/"

### CODE REVIEW

File: Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs [hunk 1]
```diff
@@ -179,4 +188,9 @@ protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage
 
     [GeneratedRegex("(?<key>\\w+)=\"(?<value>[^\"]*)\"(?:,|$)")]
     private static partial Regex BearerParameterSplitter();
+
+    private static string GetCacheKey(Uri uri)
```
Comment (BenVillalobos) [inline Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs:192]: Duplicated code here

Comment (BenVillalobos) [reply Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs:192]: It doesn't look like either this method or the GetCacheKey inside of AuthHeaderCache is called. Do we need this?

File: Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs [hunk 2]
```diff
@@ -20,6 +20,11 @@ public partial class AuthHandshakeMessageHandler : DelegatingHandler
 {
     private record AuthInfo(Uri Realm, string Service, string? Scope);
 
+    /// <summary>
+    /// Cache of most-recently-received token for each URL.
+    /// </summary>
+    private static ConcurrentDictionary<string, AuthenticationHeaderValue> HostAuthenticationCache = new();
```
Comment (BenVillalobos) [inline Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs:26]: also looks like this should be deleted

File: Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs (hunk 1 above)
Comment (rainersigwald) [reply Microsoft.NET.Build.Containers/AuthHandshakeMessageHandler.cs:192]: Yeah, I should have moved it but instead copied it. The copy in `AuthHeaderCache` is called in the other methods there.

**Review by BenVillalobos — APPROVED** [review]

**Pull request merged**